### PR TITLE
GLTF2 Exporter

### PR DESCRIPTION
**Pull request by fernandojsg**
This is a WIP GLTF v2 Exporter.

image

The code is not clean yet but I wanted to have a place to discuss this implementation from the very beginning so I hope we could end up with a fully featured exporter.

Features / TO-DO:
- [x] Include `userData` in `extras`?
- [x] **Scenes**
    - [x] Support for multiple scenes
- [ ] **Nodes**
  - [x] Meshes
    - [x] Primitive mode:
      - [x] TRIANGLES
      - [x] TRIANGLE_STRIP
      - [x] TRIANGLE_SPAN
      - [x] POINTS
      - [x] LINES
      - [x] LINE_STRIP
      - [x] LINE_LOOP
    - [x] Geometry types:
      - [x] BufferGeometry
      - [x] Geometry
    - [ ] Primitive attributes:
      - [x] POSITION
      - [x] NORMAL
      - [x] TEXCOORD_0
      - [x] TEXCOORD_1
      - [x] COLOR_0
      - [x] JOINTS_0
      - [x] WEIGHTS_0
      - [ ] TANGENT
  - [ ] Light: Need `KHR_Light` extension
  - [x] Camera
  - [ ] Skin
- [ ] **Materials**:
  - [ ] Ignore if default material is being used
  - [x] Export as lines if `material.wireframe === true`
  - [ ] `pbrMetallicRoughness` for `MeshStandardMaterial`
    - [ ] Attributes:
      - [x] `baseColorFactor`
      - [x] `metallicFactor`
      - [x] `roughnessFactor`
      - [ ] `baseColorTexture`: It's supported (`material.map`) but the `texCoord` is always set to 0.
    - [x] `doubleSided`
- [x] **Samplers**
- [ ] Images:
  - [x] `uri` using `map.image.src`
  - [ ] `uri` base64
  - [ ] `bufferView`
  - [ ] Deal with `flipY` images
- [ ] **Accessors**
  - [ ] Use the same `bufferView` for the same componentType instead of creating a new one for each attribute
  - [ ] Support `sparse`?
  - [ ] **Attributes**: 
    - [x] `bufferView`
    - [ ] `byteOffset`: Currently it's using 0 always as I'm creating a new bufferView for each accessor.
    - [x] `componentType`
    - [x] `count`
    - [x] `max`
    - [x] `min`
    - [ ] `type`:
        - [x] `SCALAR`
        - [x] `VEC2`
        - [x] `VEC3`
        - [x] `VEC4`

- [ ] **BufferViews**: Currently I'm creating a new `bufferView` for each `Accessor`, this should be fixed to use just one for these attributes that share the same `componentType`
  - [x] Attributes:
    - [x] `buffer`
    - [x] `byteOffset`
    - [x] `byteLength`
    - [x] `byteStride`
    - [x] `target`
- [x] **Buffers**: Currently I'm saving everything to a single buffer so it will be just one entry in the buffers array.
  - [x] byteLength
  - [x] uri 

Output from the current example scene: https://pastebin.com/s8Hd1AWZ

NOTE: I've added a simple example (`gltf_exporter.html`) just to have something quick to test while developing. 

/cc @donmccurdy @takahirox

### CODE REVIEW

**Comment by fernandojsg** [comment]
One of the issues I've found so far with the images is that GLTF lacks of support for `flipY`. If we're embedding the textures directly on the file we could flip them ourselves, but what if we're using url?

**Comment by fernandojsg** [comment]
In `MeshStandardMaterial` we have `metalnessMap` and `roughtnessMap` which one should we use for `metallicRoughnessTexture`? From the loader I see it's sharing the same value for both maps, but what about the opposite?
 https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/GLTF2Loader.js#L2140-L2146

**Comment by donmccurdy** [comment]
> One of the issues I've found with the images is that GLTF lacks of support for flipY. If we're embedding the textures directly on the file we could flip them ourselves, but what if we're using url?

is OBJExporter geometry-only? I guess this hasn't come up yet. Maybe support for `flipY` could be added to [a texture-related extension](https://github.com/KhronosGroup/glTF/pull/1015/files), but I would lean toward embedding the image data if it's manageable (see below).

> In MeshStandardMaterial we have metalnessMap and roughtnessMap which one should we use for metallicRoughnessTexture? From the loader I see it's sharing the same value for both maps, but what about the opposite?

Roughness and metalness must be packed together in a single texture for glTF. Also `occlusionTexture` could point to that same texture, but that is not required.

**Comment by mrdoob** [comment]
What's `examples/321103__nsstudios__blip1.wav` for? 🤔

**Comment by mrdoob** [comment]
> Geometry

I think for the time I would just do `geometry.toBufferGeometry()`.

**Comment by donmccurdy** [comment]
> KHR_Material_common

^I would omit this. Most of the extension probably won't be added to glTF 2.0, just a slimmed down Phong-only version, which isn't finished yet.

**Comment by fernandojsg** [comment]
@mrdoob 
> What's examples/321103__nsstudios__blip1.wav for? 🤔

My fault :angel: I don't know what I was doing there :)

> Include userData in extras?

Right now you can store everything you want in `userData` right? So how to store it? Serializing the json? base64? just hope that userdata is just a JSON or array and just leave it like that?

> geometry.toBufferGeometry()

Cool, I'll give it a try thanks.

**Comment by fernandojsg** [comment]
@donmccurdy 
> Roughness and metalness must be packed together in a single texture for glTF. Also occlusionTexture could point to that same texture, but that is not required.

Ok, I've just read that we could have `roughtness` in G, `metalness` in B and `occlusion` in R.

> ^I would omit this. Most of the extension probably won't be added to glTF 2.0, just a slimmed down Phong-only version, which isn't finished yet.

So what we do when the material is not `MeshStandardMaterial` ? just store a `pbr` filling the attributes we could match?

**Comment by donmccurdy** [comment]
> So what we do when the material is not MeshStandardMaterial ? just store a pbr filling the attributes we could match?

You could implement the (*DRAFT*) [KHR_materials_cmnBlinnPhong](https://github.com/UX3D-nopper/glTF/tree/master_lights_blinnphong/extensions/Khronos/KHR_materials_cmnBlinnPhong). But that spec may change before it is finalized, and it isn't supported in my viewer yet. It's not possible to represent lambert or constant materials in glTF 2.0, and probably will not be any time soon, but converting to PBR might be good enough.

**Comment by mrdoob** [comment]
>> Include userData in extras?
>
> Right now you can store everything you want in `userData` right? So how to store it? Serializing the json? base64? just hope that userdata is just a JSON or array and just leave it like that?

I think extras is supposed to be a json. So I would just `JSON.stringify()` it.

**Comment by fernandojsg** [comment]
@donmccurdy ok I'll just try to convert them to PBR and we'll handle the other cases in the next iterations.

@mrdoob cool, this is what I'm using now
```javascript
if ( object.userData ) {

  try {

    gltfNode.extras = JSON.parse( JSON.stringify( object.userData ) );

  } catch (e) {

    throw new Error( 'GLTFExporter: userData can\'t be serialized' );

  }

}
```

**Comment by fernandojsg** [comment]
@mrdoob I can't find `toBufferGeometry` is still not implemented?

**Comment by fernandojsg** [comment]
my fault, I believe you were referring to `BufferGeometry.fromGeometry`

**Comment by fernandojsg** [comment]
GLTF doesn't have support for the `material.flatShading = true` property so the exporter will save the normals as they're on the geometry (smooth by vertex) and the loader will just use them showing smooth shading instead of flat.
I believe the best way to get this working is to just compute normals per face instead of per vertex when exporting if `flatShading = true`. This means that we should duplicate vertices as each one would as many normals as face it's connected to.
What do you think @mrdoob ?

File: examples/gltf_exporter.html [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+				camera = new THREE.PerspectiveCamera( 45, window.innerWidth / window.innerHeight, 1, 2000 );
+				camera.position.set(600, 400, 0);
+
+				camera.name = "PerspectiveCamera";
+				scene1.add( camera );
+
+				// ---------------------------------------------------------------------
+				// Ambient light
+				// ---------------------------------------------------------------------
+				scene1.add( new THREE.AmbientLight( 0xffffff ) );
```
Comment (WestLangley) [inline examples/gltf_exporter.html]: I have noticed that the glTF examples are over-bright. This should be something like:
```
scene1.add( new THREE.AmbientLight( 0xffffff, 0.2 ) );
```

File: examples/gltf_exporter.html [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+
+				// ---------------------------------------------------------------------
+				// DirectLight
+				// ---------------------------------------------------------------------
+				var light = new THREE.DirectionalLight( 0xffffff );
```
Comment (WestLangley) [inline examples/gltf_exporter.html]: This is OK. Same as:
```
var light = new THREE.DirectionalLight( 0xffffff, 1 );
```

**Comment by WestLangley** [comment]
Just so everyone is on the same page, with the exception of `MeshLambertMaterial`, vertex normals are neither used, nor required, when rendering with flat shading.

**Comment by fernandojsg** [comment]
@WestLangley yes, but as GLTF doesn't allow us to define a flat shading material (AFAIK, /cc @donmccurdy) the only way we could differentiate between flat and smooth is by processing the vertex shader before exporting them

File: examples/gltf_exporter.html (hunk 1 above)
Comment (fernandojsg) [inline examples/gltf_exporter.html]: Thanks, done!

**Comment by fernandojsg** [comment]
The exporter is already fully functional, the most important parts that are missing in this PR:
- Normals for flat shading
- Textures processing:
  - Convert images to base64
  - Flip images if `flipY=true`
  - Merge `roughnessTexture` + `metalnessTexture` +  `aoTexture` 

Here is the current example exported and imported on @donmccurdy 's viewer (note that ambient light is different too)
image

image

I believe it's better to review the current features so we could keep evolving it in smaller feature-based PRs that would be easier to review than growing this one too much.

**Comment by donmccurdy** [comment]
@fernandojsg If normals are omitted, glTF requires viewers to display the mesh with flat shading ([source](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#meshes)).

**Comment by mrdoob** [comment]
> I believe the best way to get this working is to just compute normals per face instead of per vertex when exporting if `flatShading = true`. This means that we should duplicate vertices as each one would as many normals as face it's connected to.

Sounds a bit too complicated. I would ignore this for now.

**Comment by mrdoob** [comment]
Do you want to merge this for now?

**Comment by fernandojsg** [comment]
@donmccurdy cool, I didn't know that, thanks! So a first approach could be to just delete de normals for the mesh with `flatShading=true`.
@mrdoob yes please, so we could keep working on the details separately without growing the PR so much.

**Comment by mrdoob** [comment]
Many thanks!

**Pull request merged**